compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -1147,9 +1147,13 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         mut values: Option<ty::ParamEnvAnd<'tcx, ValuePairs<'tcx>>>,
         terr: TypeError<'tcx>,
         prefer_label: bool,
+        override_span: Option<Span>,
     ) {
-        let span = cause.span();
-
+        // We use `override_span` when we want the error to point at a `Span` other than
+        // `cause.span`. This is used in E0271, when a closure is passed in where the return type
+        // isn't what was expected. We want to point at the closure's return type (or expression),
+        // instead of the expression where the closure is passed as call argument.
+        let span = override_span.unwrap_or(cause.span);
```

Comment:
```suggestion
        let span = override_span.unwrap_or(cause.span());
```

Turns out that `.span` and `.span()` do different things! This is kinda what I meant when I said that these APIs are unmaintainable today.

Reply:
This should fix the regression in spans, afaict.

Reply:
https://github.com/rust-lang/rust/pull/132243 removes the need for that pointless/confusing method.